src/fs/cluster_allocator.cc

```diff
@@ -21,34 +21,71 @@
 
 #include "fs/cluster.hh"
 #include "fs/cluster_allocator.hh"
+#include "seastar/core/future.hh"
 
 #include <cassert>
 #include <optional>
 
-namespace seastar {
+namespace seastar::fs {
 
-namespace fs {
+cluster_allocator::cluster_allocator() : _cluster_sem(0) {}
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
+void cluster_allocator::init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters) {
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    assert(_cluster_sem.waiters() == 0);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
```

Comment:
something like `_cluster_sem = semaphore(_free_clusters.size())` doesn't do the job?

Reply:
Assignment operators are deleted for semaphore.

Reply:
Oh, so be it then.
Btw. I've an idea that we could assert that available_units() == _free_clusters.size() at the beginning of the function to make sure that nobody is using the allocator right now.